src/vm/lib/metadata/agent.js

```diff
@@ -574,299 +763,243 @@ function (zonename, callback) {
-         *     '<zoneroot>/foo/bar',
-         *     '<zoneroot>/foo'
-         *  ]
-         *
-         * and then attempt to ensure each component exists.
-         */
-        d = sockdir;
-        while (d.length > zoneroot.length) {
-            dirs.push(d);
-            d = path.dirname(d);
-        }
+    var zlog = self.zlog[zopts.zone];
 
-        assert.ok(dirs.length > 0, 'should have at least one dir');
+    if (!zlog) {
+        // if there's no zone-specific logger, use the global one
+        zlog = self.log;
+    }
 
-        vasync.forEachPipeline({
-            inputs: dirs,
-            func: _ensureDir
-        }, callback);
+    if (!self.zones[zopts.zone]) {
+        zlog.info({zonename: zopts.zone},
+            'zone no longer exists, not creating zsock');
+        callback();
+        return;
     }
 
-    function _attemptCreate(callback) {
-        _ensureSockpathDirsExist(function _ensureSockpathDirsExistCb(error) {
-            if (error) {
-                callback(error);
-                return;
-            }
-            zsock.createZoneSocket(zopts, callback);
-        });
+    // Ignore zones we've already got a connection for and then immediately
+    // create an entry if we don't. To act as a mutex.
+    if (self.zoneConnections[zopts.zone]) {
+        zlog.trace({zonename: zopts.zone}, 'already have zoneConnections[] for '
+            + 'zone -- not replacing.');
+        callback();
+        return;
     }
+    self.zoneConnections[zopts.zone] = {};
 
     self.addDebug(zopts.zone, 'last_zsock_create_attempt');
-    _attemptCreate(function (error, fd) {
-        var server;
-
+    _ensureSockpathDirsExist(zopts, zlog, function _ensureDirsExistCb(error) {
         if (error) {
-            zoneExists(zopts.zone, function _zoneExists(_, exists) {
-                if (!exists) {
-                    zlog.warn({err: error}, 'error creating socket and zone no '
-                        + 'longer exists, not retrying');
-                    return;
-                }
-                // If we get errors trying to create the zone socket, setup a
-                // retry loop and return.
-                zlog.error({err: error}, 'createZoneSocket error, %s seconds '
-                    + 'before next attempt', waitSecs);
-                _retryCreateZoneSocketLater();
-                return;
-            });
+            self.addDebug(zopts.zone, 'last_zsock_create_failure', error);
+
+            zlog.warn({zonename: zopts.zone, err: error},
+                'failed to create sockpath directory');
+
+            // We were unable to create the directory but we have not yet
+            // created a real self.zoneConnections entry so we can just delete
+            // the placeholder and call callback. If the VM still exists and
+            // is running, we'll try again when we next poll w/
+            // _checkNewZones().
+            delete self.zoneConnections[zopts.zone];
+            callback();
```

Comment:
It seems we're swallowing the error `error` here, but that at least some `createZoneSocket` call sites use the error passed in its callback, so there might be some inconsistency.